Reactions don't register if message already registered.

**Describe the bug**
Even though the bot says...
> [ERROR] Registering button ⏮️failed. Message already registered.

it should still cancel that registration and attempt to re-register. If the bot doesn't, the program will not listen to the right on_reaction_add event.

--- a/cogs/event_listener.py
+++ b/cogs/event_listener.py
@@ -7,6 +7,7 @@ from src.parser import parse
 from src.command_handler import play_ytid
 import src.element.profile
 from discord.ext import commands
+from discord_components import DiscordComponents
 from src.constants import INVITE_LINK_FORMAT
 
 COMMAND_CHANNEL_WARNING = 'Accepted command.'
@@ -102,6 +103,7 @@ class EventListener(commands.Cog):
 
     @commands.Cog.listener()
     async def on_ready(self):
+        DiscordComponents(self.bot)
         await src.preinitialization.generate_profiles(bot.guilds, self.bot, profiles)
         for profile in profiles:
             await profile.setup()

--- a/requirements.txt
+++ b/requirements.txt
@@ -1,4 +1,4 @@
-discord-py-slash-command>=1.1.2
+discord-py-slash-command==1.2.2
 discord.py>=1.6.0
 requests>=2.25.0
 youtube-search>=1.1.1
@@ -7,3 +7,4 @@ PyNaCl==1.4.0
 youtube-dl==2021.4.17
 python-dotenv==0.17.1
 autopep8==1.5.7
+discord-components==0.5.3

--- a/src/commander/element/EventWatcher.py
+++ b/src/commander/element/EventWatcher.py
@@ -16,27 +16,27 @@ class EventWatcher:
         self.coro: Dict[str, asyncio.Task] = {}
     
     async def watch_button_click(self, emoji: str, timeout = None, action: FunctionType = None, action_timeout: FunctionType = None):
-        if emoji not in self.coro:
+        if emoji in self.coro:
+            logging.warn(f'button_click: {emoji} event is already registered. Overwriting event and registering again.')
+            self.remove(emoji=emoji)
+            
+        async def refresh():
+            def check(interaction: Interaction):
+                return interaction.component.emoji.name == emoji
 
-            async def refresh():
-                def check(interaction: Interaction):
-                    return interaction.component.emoji.name == emoji
+            try:
+                interaction: Interaction = await self.client.wait_for('button_click', timeout=timeout, check=check)
+            except asyncio.TimeoutError:
+                logging.info(emoji + ' button_click event timed out')
+                await self.exec_and_await(action_timeout)
+            else:
+                logging.debug(emoji + ' button_click event pressed')
+                self.coro[emoji] = asyncio.create_task(refresh())
+                await self.exec_and_await(action)
+                await interaction.respond(type=6)
+    
+        self.coro[emoji] = asyncio.create_task(refresh())
 
-                try:
-                    interaction: Interaction = await self.client.wait_for('button_click', timeout=timeout, check=check)
-                except asyncio.TimeoutError:
-                    logging.info(emoji + ' button_click event timed out')
-                    await self.exec_and_await(action_timeout)
-                else:
-                    logging.debug(emoji + ' button_click event pressed')
-                    self.coro[emoji] = asyncio.create_task(refresh())
-                    await self.exec_and_await(action)
-                    await interaction.respond(type=6)
-        
-            self.coro[emoji] = asyncio.create_task(refresh())
-
-        else:
-            logging.error(f'button_click: {emoji} event is already registered.')
     
     async def exec_and_await(self, function):
         res = function()

--- a/src/commander/element/EventWatcher.py
+++ b/src/commander/element/EventWatcher.py
@@ -1,0 +1,62 @@
+import discord
+import asyncio
+import logging
+import inspect
+from types import FunctionType
+from typing import Dict
+from discord import emoji
+from discord_components import Interaction
+
+class EventWatcher:
+    """
+    Handles on press events and actions.
+    """
+    def __init__(self, client: discord.Client):
+        self.client = client
+        self.coro: Dict[str, asyncio.Task] = {}
+    
+    async def watch_button_click(self, emoji: str, timeout = None, action: FunctionType = None, action_timeout: FunctionType = None):
+        if emoji not in self.coro:
+
+            async def refresh():
+                def check(interaction: Interaction):
+                    return interaction.component.emoji.name == emoji
+
+                try:
+                    interaction: Interaction = await self.client.wait_for('button_click', timeout=timeout, check=check)
+                except asyncio.TimeoutError:
+                    logging.info(emoji + ' button_click event timed out')
+                    await self.exec_and_await(action_timeout)
+                else:
+                    logging.debug(emoji + ' button_click event pressed')
+                    self.coro[emoji] = asyncio.create_task(refresh())
+                    await self.exec_and_await(action)
+                    await interaction.respond(type=6)
+        
+            self.coro[emoji] = asyncio.create_task(refresh())
+
+        else:
+            logging.error(f'button_click: {emoji} event is already registered.')
+    
+    async def exec_and_await(self, function):
+        res = function()
+        if inspect.isawaitable(res):
+            await res
+    
+    async def remove(self, emoji: str):
+        """Removes this event from the eventwatcher.
+        
+        This function cancels the coroutine."""
+        if self.coro[emoji]:
+            self.coro[emoji].cancel()
+            del self.coro[emoji]
+        else:
+            logging.error('Label does not exist.')
+
+    async def remove_all(self):
+        """Removes all events from the eventwatcher.
+        
+        This function cancels the coroutine."""
+        for key in list(self.coro):
+            self.coro[key].cancel()
+            del self.coro[key]

--- a/src/music/player.py
+++ b/src/music/player.py
@@ -187,11 +187,8 @@ class Player:
         self.default_queue_metadata()
         self.default_player_metadata()
 
-        # Function is ran in async (using run_coroutine_threadsafe as opposed to being awaited)
-        # so that buttons can register while the player is free to do other things.
-        # Be aware that this might mask errors in that function
         if self.player_message:
-            asyncio.run_coroutine_threadsafe(self.register_player_reactions(self.player_message), self.client.loop)
+            await self.register_player_reactions(self.player_message)
 
     async def disconnect(self):
         """Disconnects the player to a voicechannel"""
@@ -207,11 +204,8 @@ class Player:
         self.default_queue_metadata()
         self.default_player_metadata()
 
-        # Function is ran in async (using run_coroutine_threadsafe as opposed to being awaited)
-        # so that buttons can register while the player is free to do other things.
-        # Be aware that this might mask errors in that function
         if self.player_message:
-            asyncio.run_coroutine_threadsafe(self.remove_player_reactions(), self.client.loop)
+            await self.remove_player_reactions()
 
     ########## MusicSource Event Handlers ##########
     def on_finished(self, error):

--- a/src/music/player.py
+++ b/src/music/player.py
@@ -1,16 +1,18 @@
 import threading
-from typing import Dict
+from typing import Dict, List
 from discord.channel import TextChannel, VoiceChannel
 from discord.client import Client
 from discord.message import Message
 from discord.player import FFmpegPCMAudio
 from discord.voice_client import VoiceClient
 from datetime import timedelta
+from discord_components.button import Button
 import youtube_dl
 import logging
 import asyncio
 
 from src.commander.element.Reaction import Reaction
+from src.commander.element.EventWatcher import EventWatcher
 from src.commander.EmbedFactory import EmbedFactory
 from src.commander.element.ReporterEmbed import ReporterEmbed
 from src.commander.element.LyricsEmbed import LyricsEmbed
@@ -47,13 +49,15 @@ class Player:
         # Footer Metadata
         self.volume: int= 50
 
+        self.button_events = EventWatcher(self.client)
+
         self.player_reactions = [
             Reaction(emoji='⏮️', client=self.client, action=self.last),
             Reaction(emoji='⏯️', client=self.client, action=self.on_play_pause),
             Reaction(emoji='⏭️', client=self.client, action=self.next),
             Reaction(emoji='🔉', client=self.client, action=self.lower_volume),
             Reaction(emoji='🔊', client=self.client, action=self.raise_volume),
-            Reaction(emoji='💬', client=self.client, action=self.toggle_display_description)
+            # Reaction(emoji='💬', client=self.client, action=self.toggle_display_description)
         ]
 
         # Front End Registration Objects
@@ -353,6 +357,15 @@ class Player:
     
     ########## Player ##########
     @property
+    def player_buttons(self) -> List[Button]:
+        button_format: list = []
+        for reaction in self.player_reactions:
+            if reaction.emoji in self.button_events.coro:
+                button_format.append(Button(emoji=reaction.emoji))
+        
+        return [button_format] if button_format else []
+
+    @property
     def player_embed(self) -> PlayerEmbed:
         """Getter for player_embed from factory.
         Factory generates a up to date embed based on kwargs passed.
@@ -363,7 +376,7 @@ class Player:
     async def update_player_embed(self) -> None:
         """Edits the player message (if registered) with a newly generated embed."""
         if self.player_message:
-            await self.player_message.edit(embed = self.player_embed)
+            await self.player_message.edit(embed = self.player_embed, components=self.player_buttons)
     
     def default_player_metadata(self) -> None:
         """Sets the player metadata variables to default values.
@@ -391,16 +404,13 @@ class Player:
     
     ########## Reaction Functions ##########
     async def register_player_reactions(self, message: Message):
-        """Register all the reactions in this class"""
-        if self.player_reactions:
-            for reactions in self.player_reactions:
-                await reactions.register(message)
+        for reaction in self.player_reactions:
+            await self.button_events.watch_button_click(emoji=reaction.emoji, action= reaction.action)
 
     async def remove_player_reactions(self):
         """Removes all the reactions in this class"""
-        if self.player_reactions:
-            for reactions in self.player_reactions:
-                await reactions.remove_all()
+        await self.button_events.remove_all()
+        await self.update_player_embed()
 
     async def toggle_display_description(self):
         """Intake function for 💬 emoji reaction on_press event"""

